sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_azureappconfigurationprovider.py

```diff
@@ -261,18 +264,194 @@ def _is_json_content_type(content_type: str) -> bool:
     settings from Azure App Configuration into a Python application. Enables trimming of prefixes from configuration
     keys. Enables resolution of Key Vault references in configuration settings.
     """
 
-    def __init__(self) -> None:
+    def __init__(self, **kwargs) -> None:
         self._dict: Dict[str, str] = {}
         self._trim_prefixes: List[str] = []
         self._client: Optional[AzureAppConfigurationClient] = None
         self._secret_clients: Dict[str, SecretClient] = {}
+        self._key_vault_options: Optional[AzureAppConfigurationKeyVaultOptions] = kwargs.pop("key_vault_options", None)
+        self._selects: List[SettingSelector] = kwargs.pop(
+            "selects", [SettingSelector(key_filter="*", label_filter=EMPTY_LABEL)]
+        )
+
+        trim_prefixes: List[str] = kwargs.pop("trim_prefixes", [])
+        self._trim_prefixes = sorted(trim_prefixes, key=len, reverse=True)
+
+        refresh_on: List[Tuple[str, str]] = kwargs.pop("refresh_on", None) or []
+        self._refresh_on: Mapping[Tuple[str, str] : Optional[str]] = {_build_sentinel(s): None for s in refresh_on}
+        self._refresh_timer: _RefreshTimer = _RefreshTimer(**kwargs)
+        self._on_refresh_error: Optional[Callable[[Exception], None]] = kwargs.pop("on_refresh_error", None)
+        self._update_lock = Lock()
+
+    def refresh(self, **kwargs) -> None:
+        if not self._refresh_on:
+            logging.debug("Refresh called but no refresh options set.")
+            return
+
+        if not self._refresh_timer.needs_refresh():
+            logging.debug("Refresh called but refresh interval not elapsed.")
+            return
+        try:
+            with self._update_lock:
+                for (key, label), etag in self._refresh_on.items():
+                    updated_sentinel = self._client.get_configuration_setting(
+                        key=key, label=label, etag=etag, match_condition=MatchConditions.IfModified, **kwargs
+                    )
+                    if updated_sentinel is not None:
+                        logging.debug(
+                            "Refresh all triggered by key: %s label %s.",
+                            key,
+                            label,
+                        )
+                        self._load_all(**kwargs)
+                        self._refresh_on[(key, label)] = updated_sentinel.etag
+                        self._refresh_timer.reset()
+                        return
+        except (ServiceRequestError, ServiceResponseError) as e:
+            logging.debug("Failed to refresh, retrying: %r", e)
+            self._refresh_timer.retry()
+        except HttpResponseError as e:
+            # There might be specific status codes that we want to silently backoff and retry.
+            # Need more service specific details here, raising for now.
+            if _is_retryable_error(e):
```

Comment:
Why are we adding another layer of retries in the provider? SDK already retries on the errors listed here. We can just update the next refresh time and let the next call to refresh take care of retries.

Reply:
i.e. we update the next refresh time on all refresh failures (not just retriable error codes)

Reply:
Any error that is not specifically expected (e.g. the 429, or network instability like ServiceResponseError) should be raised, rather than retried - hence the addition of the `_is_retryable_error`.
If we feel the extra layer of retries is superfluous over and above what the SDK already does - the alternative would be to raise everything.

Reply:
I think this may have been muddled at some point. We should always do the backoff retry, not just on a few errors

Reply:
@annatisch, we aren't stopping the error from being raised. What this is doing is saying the we will not wait the full refresh interval for the next attempt at a refresh.